### ISSUE TITLE
Why does whisker disappear in the karyotype figure of 3 species?

### ISSUE THREAD

**Issue by sophy7074** [tanghaibao/jcvi#23]
Hi HaiBao,
I got the karyotype figure successfully following https://github.com/tanghaibao/jcvi/wiki/MCscan-(Python-version).
Firstly, I got the karyotype figure between A and B.
![image](https://cloud.githubusercontent.com/assets/8351253/18045645/6b7598d4-6e06-11e6-94ab-c6d5a15fb6b3.png)

Secondly, I got the karyotype figure between B and C. 
![image](https://cloud.githubusercontent.com/assets/8351253/18045633/582bdc8e-6e06-11e6-8057-90d6c03a1c1b.png)

Then I `cat B.A.anchors B.C.anchors > B.2Spp.anchors;
cat A.bed C.bed > 2Spp.bed;
python -m jcvi.compara.synteny screen --minspan=10 --simple B.2Sp.anchors B.2Spp.anchors.simple` and modified the context of seqids and layout:
**seqids:**
ChrA1,ChrA2,ChrA3,...
scaffoldB1,scaffoldB2,scaffoldB3,...
chC1,chC2,chC3,...
**layout:**
\# y, xstart, xend, rotation, color, label,     va,      bed
 .8,     .1,    .8,       0,  blue,  A,      top,  2Spp.bed
 .6,     .1,    .8,       0,   red,  B, bottom,       B.bed
 .4,     .1,    .8,       0, green, C,    top,    2Spp.bed
\# edges
e, 0, 1, B.2Spp.anchors.simple

but the whisker disappeared in figure:
![image](https://cloud.githubusercontent.com/assets/8351253/18045439/6900acfc-6e05-11e6-8aae-1d905fec3995.png)

How can I add the whisker?

Sincerely,
Tiany

**Comment by tanghaibao** [comment]
You don't have to cat. Try something like below. "e, 0, 1" means draw whiskers between track-0 (top track at y=.8) and track-1 (track at y=.6). "e, 1, 2" means draw whiskers between track-1 and track-2 (track at y=.4).

```
# y, xstart, xend, rotation, color, label, va, bed
.8, .1, .8, 0, blue, A, top, A.bed
.6, .1, .8, 0, red, B, bottom, B.bed
.4, .1, .8, 0, green, C, top, C.bed
# edges
e, 0, 1, A.B.anchors.simple
e, 1, 2, B.C.anchors.simple
```

**Comment by ascendo** [comment]
Hi HaiBao,

The karyotype size is somewhat different from real scaffold.

I think the end of scaffolds have no gene in there.

**Comment by tanghaibao** [comment]
The default size corresponds to the number of genes, not base pairs. It is possible to draw sizes that match bp size with some code modifications though.

**Comment by ascendo** [comment]
Oh I see. Thank you .

**Comment by sophy7074** [comment]
Hi HaiBao,

Thank you! I got the picture that I want following your advise.

When I highlighted a particular synteny block: r*, there is no red line in picture. Is this because the small size of block which span 20 genes? 
Before highlight:
![1](https://cloud.githubusercontent.com/assets/8351253/18260848/4d0631be-7424-11e6-8d4d-a4ff6836d14f.PNG)
After highlight:
![2](https://cloud.githubusercontent.com/assets/8351253/18260982/96aded38-7425-11e6-9dfc-4147457e0b38.PNG)
Nothing was changed. Red line did not appear as I expectated.
![3](https://cloud.githubusercontent.com/assets/8351253/18261017/d90ee826-7425-11e6-93e4-2bb056dfca25.PNG)

Another question is how to draw whiskers start at the top of scaffold and end at the top of scaffold. For example, I thinks it better when the whisker between blue and green are on the top of scaffold in the figure below. 
![qq 20160906104010](https://cloud.githubusercontent.com/assets/8351253/18260296/d4afba50-741e-11e6-8954-f98f55f41f42.png)

Best wishes!
Tiany

**Comment by tanghaibao** [comment]
The highlight should appear even if the selected block is small. I'm not sure what happened there. For the second question, plot the whiskers 'above' is possible, with related code (``samearc`` parameter) is here:
https://github.com/tanghaibao/jcvi/blob/master/graphics/karyotype.py#L214

I need to make some code fixes to make that work. Would you send me your config files so that I could test the change?

**Comment by sophy7074** [comment]
Hi HaiBao,

When I processed data and config files, I found the first question happend because the scaffold where blocks locate were deleted in seqids. There is nothing wrong about highlight in jcvi.

About the second problem, it didn't work after I change samearc parameter.
Line 124 `samearc = "above" if i == j and i == 0 else "below"` in karyotype.py means that whiskers only can be drawn above when they connect scaffolds in same specie (i = j).  In my figure, they connect scaffolds from different species. If I change edges in layout file, a KeyError occurred because it can not find track-1.

Besides, I read the code after KeyError occurred. I found that the code https://github.com/tanghaibao/jcvi/blob/master/compara/synteny.py#L238 to https://github.com/tanghaibao/jcvi/blob/master/compara/synteny.py#L241 make no sense.

If genes in blocks exists in *.simple but not in *.bed, the order of genes in block exchange (Like the first problem). Then https://github.com/tanghaibao/jcvi/blob/master/graphics/karyotype.py#L221 fails because a and b from `track[j]` can not exist in atrack (`track[i]`).
Example:
KeyError:
![3](https://cloud.githubusercontent.com/assets/8351253/18303226/93b39404-750e-11e6-9da1-d081cfa66031.PNG)
Block:
![1](https://cloud.githubusercontent.com/assets/8351253/18303170/502aad44-750e-11e6-8030-7241630d9a0d.PNG)
Bed:
![2](https://cloud.githubusercontent.com/assets/8351253/18303175/56217f5c-750e-11e6-8dad-48b65b68af4e.PNG)
In fact, the gene `MAG11416-RA` did not exist in Mage.bed. It is difficult to correct.

If the bed file can't be loaded seem as the situation in second problem after I change `e, 0, 0, *.simple`, or everthing is right, these codes don't even run.
I modify the code:
![image](https://cloud.githubusercontent.com/assets/8351253/18303686/ea67ac2a-7510-11e6-9e9f-e185a3cd783c.png)

Files I used for plotting are in https://github.com/sophy7074/jcvi-karotype-example .

Thank you!
Tiany

**Comment by tanghaibao** [comment]
Tiany,

Good that you find the issues. I pointed you the samearc parameter, but
there needs to be some code changes. If your top two tracks are track-0 and
track-1 for example, you can set samearc = "above" if (i, j) == (0, 1). A
cleaner solution is to put the samearc option in the edge lines.

The block exchange code is there to handle the situation when the query and
subject are switched. For example, your .simple file may have tomato in the
query (first two columns), but instead should be in the subject (next two
columns). The code will then effectively check existence of gene in both
query and subject. In your case, I don't yet understand why MAG11416-RA
doesn't exist in the .bed since the .simple file should be derived from the
synteny calculation which uses the .bed file.

Sorry for the confusion. Some behavior of the code requires a good
understanding of the peculiarities of the synteny blocks. If you find a
solution, please make a pull request and I'll merge your changes later.

Haibao

**Comment by sophy7074** [comment]
Hi Haibao,

The method that set samearc = "above" if (i, j) == (0, 1) works only when plot track-0 and track-1.
I'm favor the second way. So I changed some codes in karyotype.py to make it true that draw whiskers above or below according users' setting. Differences in script are that adding line 70-73, modifing line 77, line 215 and commenting line 217 out. The layout file can set as below, all 4 cases are right:
![1](https://cloud.githubusercontent.com/assets/8351253/18340919/2ad14482-75da-11e6-968a-22e6f54097d3.PNG)
I got the figure that I want:
![2](https://cloud.githubusercontent.com/assets/8351253/18341128/1b4cbc02-75db-11e6-8d62-84a305f827c2.PNG)

The reason for MAG11416-RA doesn't exist in the .bed but in the .simple file is that I selected part of scaffolds to draw but didn't select blocks in simple files. Sorry for causing your trouble.

About the block exchange code, I understand that query and subject need be switched in some situation. However, I thinks it necessary to make sure whether genes from subject are in order after switch. It will continue to fail if not. So I add a condition to judge whether c in order or not and reminder user to correct wrong files.

My problems are solved now. 

Thank you,
Tiany
